File: Samples/Basic/DocumentReaderObjectiveC-sample/DocumentReaderObjectiveC-sample/ViewController.m
//
//  ViewController.m
//  DocumentReaderObjectiveC-sample
//
//  Created by Dmitry Smolyakov on 3/19/18.
//  Copyright © 2018 Dmitry Smolyakov. All rights reserved.
//

#import "ViewController.h"
#import "RGRecognizeImageViewController.h"

@import DocumentReader;
@import Photos;

@interface ViewController () <UIPickerViewDelegate, UIPickerViewDataSource,
UIImagePickerControllerDelegate, UINavigationControllerDelegate, RGRecognizeImageViewController>

@property (weak, nonatomic) IBOutlet UILabel *nameLabel;
@property (weak, nonatomic) IBOutlet UIImageView *documentImage;
@property (weak, nonatomic) IBOutlet UIImageView *portraitImageView;
@property (weak, nonatomic) IBOutlet UIPickerView *pickerView;
@property (weak, nonatomic) IBOutlet UIButton *userRecognizeImage;
@property (weak, nonatomic) IBOutlet UIButton *useCameraViewControllerButton;
@property (weak, nonatomic) IBOutlet UIButton *customCamera;

@property (weak, nonatomic) IBOutlet UILabel *initializationLabel;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;

@property (strong, nonatomic) UIImagePickerController *imagePicker;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self initializationReader];
    self.imagePicker = [[UIImagePickerController alloc] init];
    // Do any additional setup after loading the view, typically from a nib.
}

- (void)initializationReader {
    //initialize license
    NSString *dataPath = [[NSBundle mainBundle] pathForResource:@"regula.license" ofType:nil];
    NSData *licenseData = [NSData dataWithContentsOfFile:dataPath];
    
    [RGLDocReader.shared prepareDatabase:@"Full" progressHandler:^(NSProgress * _Nonnull progress) {
        self.initializationLabel.text = [NSString stringWithFormat:@"%.1f", progress.fractionCompleted * 100];
    } completion:^(BOOL successful, NSError * _Nullable error) {
        if (successful) {
            self.initializationLabel.text = @"Initialization...";
            [RGLDocReader.shared initializeReader:licenseData completion:^(BOOL successful, NSError * _Nullable error ) {
                if (successful) {
                    [self.activityIndicator stopAnimating];
                    [self.initializationLabel setHidden:YES];
                    [self.userRecognizeImage setHidden:NO];
                    [self.useCameraViewControllerButton setHidden:NO];
                    [self.customCamera setHidden:NO];
                    [self.pickerView setHidden:NO];
                    [self.pickerView reloadAllComponents];
                    [self.pickerView selectRow:0 inComponent:0 animated:NO];

                    RGLScenario *scenario = [RGLDocReader shared].availableScenarios.firstObject;
                    if (scenario) {
                        [RGLDocReader shared].processParams.scenario = scenario.identifier;
                    }
                    [RGLDocReader shared].functionality.singleResult = YES;
                    
                    for (RGLScenario *scenario in RGLDocReader.shared.availableScenarios) {
                        NSLog(@"%@", scenario);
                        NSLog(@"---------");
                    }
                } else {
                    [self.activityIndicator stopAnimating];
                    self.initializationLabel.text = [NSString stringWithFormat:@"Initialization error: %@", error];
                    NSLog(@"%@", error);
                }
            }];
        } else {
            self.initializationLabel.text = [NSString stringWithFormat:@"Downloading database error: %@", error];
            NSLog(@"%@", error);
        }
    }];
}

- (IBAction)useCameraViewController:(UIButton *)sender {
    [RGLDocReader.shared showScanner:self completion:^(enum RGLDocReaderAction action, RGLDocumentReaderResults * _Nullable result, NSError * _Nullable error) {
        switch (action) {
            case RGLDocReaderActionCancel: {
                NSLog(@"Cancelled by user");
            }
            break;

            case RGLDocReaderActionComplete: {
                NSLog(@"Completed");
                [self handleScanResults:result];
            }
            break;

            case RGLDocReaderActionError: {
                NSLog(@"Error string: %@", error);
            }
            break;

            case RGLDocReaderActionProcess: {
                NSLog(@"Scaning not finished. Result: %@", result);
            }
            break;

            default:
            break;
        }
    }];
}

- (IBAction)useRecognizeImageMethod:(UIButton *)sender {
    [self getImageFromGallery];
}

- (IBAction)actionCustomCamera:(UIButton *)sender {
    RGRecognizeImageViewController *vc = [[RGRecognizeImageViewController alloc] init];
    vc.delegate = self;
    vc.modalPresentationStyle = UIModalPresentationFullScreen;
    [self presentViewController:vc animated:YES completion:nil];
}

- (void)getImageFromGallery {
    [PHPhotoLibrary requestAuthorization:^(PHAuthorizationStatus status) {
        switch (status) {
            case PHAuthorizationStatusAuthorized: {
                if ([UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeSavedPhotosAlbum]) {
                    self.imagePicker.delegate = self;
                    self.imagePicker.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
                    self.imagePicker.allowsEditing = NO;
                    dispatch_async(dispatch_get_main_queue(), ^{
                        self.imagePicker.navigationBar.tintColor = [UIColor blackColor];
                        [self presentViewController:self.imagePicker animated:YES completion:nil];
                    });
                }
            }
            break;

            case PHAuthorizationStatusDenied: {
                NSString *message = @"Application doesn't have permission to use the camera, please change privacy settings";
                UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Gallery Unavailable" message:message preferredStyle:UIAlertControllerStyleAlert];
                [alertController addAction: [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
                [alertController addAction:[UIAlertAction actionWithTitle:@"Settings" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                    [UIApplication.sharedApplication openURL:[NSURL URLWithString:UIApplicationOpenSettingsURLString]];
                }]];
                [self presentViewController:alertController animated:YES completion:nil];
            }
            break;

            case PHAuthorizationStatusNotDetermined: {
                NSLog(@"%@", @"PHPhotoLibrary status: notDetermined");
            }

            case PHAuthorizationStatusRestricted: {
                NSLog(@"%@", @"PHPhotoLibrary status: restricted");
            }

            default:
            break;
        }
    }];
}

- (void)handleScanResults:(RGLDocumentReaderResults *)result {
    // use fast getValue method
    NSString *name = [result getTextFieldValueByType:RGLFieldTypeFt_Surname_And_Given_Names];
    NSLog(@"%@", name);
    self.nameLabel.text = name;
    self.documentImage.image = [result getGraphicFieldImageByType:RGLGraphicFieldTypeGf_DocumentImage source:RGLResultTypeRawImage];
    self.portraitImageView.image = [result getGraphicFieldImageByType:RGLGraphicFieldTypeGf_Portrait];

    for (RGLDocumentReaderTextField *textField in result.textResult.fields) {
        NSString *value = [result getTextFieldValueByType:textField.fieldType lcid:textField.lcid];
        NSLog(@"Field type name: %@, value: %@", textField.fieldName, value);
    }
}

- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info {
    UIImage *image = info[UIImagePickerControllerOriginalImage];
    [self dismissViewControllerAnimated:YES completion:^{

        [RGLDocReader.shared recognizeImage:image cameraMode:NO completion:^(RGLDocReaderAction action, RGLDocumentReaderResults * _Nullable results, NSError * _Nullable error) {
            if (action == RGLDocReaderActionComplete) {
                if (results != nil) {
                    NSLog(@"Completed");
                    [self handleScanResults:results];
                }
            } else if (action == RGLDocReaderActionError) {
                [self dismissViewControllerAnimated:YES completion:nil];
                NSLog(@"Something went wrong");
            }
        }];
    }];
}

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return RGLDocReader.shared.availableScenarios.count;
}


- (nullable NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    return RGLDocReader.shared.availableScenarios[row].identifier;
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    RGLDocReader.shared.processParams.scenario = RGLDocReader.shared.availableScenarios[row].identifier;
}

- (void)recognizeDidFinishedWith:(RGLDocumentReaderResults *)results viewController:(RGRecognizeImageViewController *)viewController {
    [self handleScanResults:results];
}

@end

File: Samples/DocumentReaderSwift-sample/DocumentReaderSwift-sample/ViewController.swift
//
//  ViewController.swift
//  DocumentReaderSwift-sample
//
//  Created by Dmitry Smolyakov on 6/13/17.
//  Copyright © 2017 Dmitry Smolyakov. All rights reserved.
//

import UIKit
import DocumentReader
import Photos

class ViewController: UIViewController {

    @IBOutlet weak var nameLabel: UILabel!
    @IBOutlet weak var documentImage: UIImageView!
    @IBOutlet weak var portraitImageView: UIImageView!

    @IBOutlet weak var pickerView: UIPickerView!
    @IBOutlet weak var userRecognizeImage: UIButton!
    @IBOutlet weak var useCameraViewControllerButton: UIButton!

    @IBOutlet weak var initializationLabel: UILabel!
    @IBOutlet weak var activityIndicator: UIActivityIndicatorView!

    var imagePicker = UIImagePickerController()

    var docReader: DocReader?

    override func viewDidLoad() {
        super.viewDidLoad()
        initializationReader()
    }

    func initializationReader() {
        //initialize license
        guard let dataPath = Bundle.main.path(forResource: "regula.license", ofType: nil) else { return }
        guard let licenseData = try? Data(contentsOf: URL(fileURLWithPath: dataPath)) else { return }

        //create DocReader object
        let docReader = DocReader()

        docReader.initilizeReader(license: licenseData) { (successfull, error) in
            if successfull {
                self.activityIndicator.stopAnimating()
                self.initializationLabel.isHidden = true
                self.userRecognizeImage.isHidden = false
                self.useCameraViewControllerButton.isHidden = false
                self.pickerView.isHidden = false
                self.pickerView.reloadAllComponents()
                self.pickerView.selectRow(0, inComponent: 0, animated: false)

                //Get available scenarios
                for scenario in docReader.availableScenarios {
                    print(scenario)
                    print("--------")
                }
            } else {
                self.activityIndicator.stopAnimating()
                let licenseError = error ?? "Unknown error"
                self.initializationLabel.text = "Initialization error: \(licenseError)"
                print(licenseError)
            }
        }

        //set scenario
        docReader.processParams.scenario = "Mrz"
        self.docReader = docReader
    }

    // Use this code for recognize on photo from camera
    @IBAction func useCameraViewController(_ sender: UIButton) {
        //start recognize
        docReader?.showScanner(self) { (action, result, error) in
            switch action {
            case .cancel:
                print("Cancelled by user")
            case .complete:
                print("Completed")
                self.handleResult(result: result)
            case .error:
                print("Error")
                guard let error = error else { return }
                print("Error string: \(error)")
            case .process:
                guard let result = result else { return }
                print("Scaning not finished. Result: \(result)")
            case .morePagesAvailable:
                print("This status couldn't be here, it uses for -recognizeImage function")
            }
        }
    }

    func handleResult(result: DocumentReaderResults?) {
        guard let result = result else { return }
        print("Result class: \(result)")
        // use fast getValue method
        let name = result.getTextFieldValueByType(fieldType: .ft_Surname_And_Given_Names)
        print("NAME: \(name ?? "empty field")")
        self.nameLabel.text = name
        self.documentImage.image = result.getGraphicFieldImageByType(fieldType: .gf_DocumentFront, source: .rawImage)
        self.portraitImageView.image = result.getGraphicFieldImageByType(fieldType: .gf_Portrait)

        //go though all text results
        for textField in result.textResult.fields {
            guard let value = result.getTextFieldValueByType(fieldType: textField.fieldType, lcid: textField.lcid) else { continue }
            print("Field type name: \(textField.fieldName), value: \(value)")
        }
    }

    // Use this code for recognize on photo from gallery
    @IBAction func useRecognizeImageMethod(_ sender: UIButton) {
        //load image from assets folder
        getImageFromGallery()
    }

    func getImageFromGallery() {
        PHPhotoLibrary .requestAuthorization { (status) in
            switch status {
            case .authorized:
                if UIImagePickerController.isSourceTypeAvailable(.savedPhotosAlbum){
                    self.imagePicker.delegate = self
                    self.imagePicker.sourceType = .photoLibrary;
                    self.imagePicker.allowsEditing = false
                    DispatchQueue.main.async {
                        self.imagePicker.navigationBar.tintColor = .black
                        self.present(self.imagePicker, animated: true, completion: nil)
                    }
                }
            case .denied:
                let message = NSLocalizedString("Application doesn't have permission to use the camera, please change privacy settings", comment: "Alert message when the user has denied access to the gallery")
                let alertController = UIAlertController(title: NSLocalizedString("Gallery Unavailable", comment: "Alert eror title"), message: message, preferredStyle: .alert)
                alertController.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: "Alert manager, OK button tittle"), style: .cancel, handler: nil))
                alertController.addAction(UIAlertAction(title: NSLocalizedString("Settings", comment: "Alert button to open Settings"), style: .default, handler: { action in
                    if #available(iOS 10.0, *) {
                        guard let settingsURL = URL(string: UIApplicationOpenSettingsURLString) else { return }
                        UIApplication.shared.open(settingsURL, options: [:], completionHandler: nil)
                    } else {
                        UIApplication.shared.openURL(URL(string: UIApplicationOpenSettingsURLString)!)
                    }
                }))
                self.present(alertController, animated: true, completion: nil)
                print("PHPhotoLibrary status: denied")
                break
            case .notDetermined:
                print("PHPhotoLibrary status: notDetermined")
            case .restricted:
                print("PHPhotoLibrary status: restricted")
            }
        }
    }

}

extension ViewController: UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {
        if let image = info[UIImagePickerControllerOriginalImage] as? UIImage {
            self.dismiss(animated: true, completion: {

                //start recognize
                self.docReader?.recognizeImage(image, completion: { (action, result, error) in
                    if action == .complete {
                        if result != nil {
                            print("Completed")
                            print("Result class: \(result!)")
                            self.handleResult(result: result)
                        } else {
                            print("Completed without result")
                        }
                    } else if action == .error {
                        print("Eror")
                        guard let error = error else { return }
                        print("Eror: \(error)")
                    }
                })

            })
        } else {
            self.dismiss(animated: true, completion: nil)
            print("Something went wrong")
        }
    }
}

extension ViewController: UIPickerViewDataSource {
    public func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }

    public func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        guard let docReader = docReader else { return 0 }
        return docReader.availableScenarios.count
    }
}

extension ViewController: UIPickerViewDelegate {
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return docReader?.availableScenarios[row].identifier
    }

    public func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        guard let docReader = docReader else { return }
        self.docReader?.processParams.scenario = docReader.availableScenarios[row].identifier
    }
}

